Typing Indicator Support

nathantannar4 opened the pull request:
This PR adds TypingIndicator support by reserving the last section of the MessageCollectionView as needed. MessageKit will support the iMessage style typing bubble out of the box but you can also implement your own typing indicator cells!

nathantannar4 committed:
Working Prototype

nathantannar4 committed:
Changes suggested by SD10

nathantannar4 committed:
TypingIndicator API adjustments

nathantannar4 committed:
Circle ->  BubbleCircle

nathantannar4 committed:
Public access, needed for custom FlowLayouts, custom cells, etc

nathantannar4 committed:
API adjustments to choose TypingBubble colors with properties

nathantannar4 committed:
Add CHANGELOG, rename TypingBubble -> TypingIndicator

zhongwuzw commented:
When the indicator is animating, the same time I post a message, indicator cell would scroll to the bottom, this scroll animation is what we want?

nathantannar4 commented:
I think that's how other messaging apps handle it

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

SD10 commented:
I'll have to check this animation you guys are talking about. This is a feature that MessageKit will support, but personally I will never use it

nathantannar4 committed:
Add default backgroundColor

nathantannar4 committed:
Smoother Animations, more like iMessage

SD10 commented on Example/Sources/ConversationViewController.swift:
```diff
@@ -82,28 +82,32 @@ internal class ConversationViewController: MessagesViewController {
+        setTypingIndicatorHidden(!isTyping, animated: true, completion: { _ in
+            self.messagesCollectionView.scrollToBottom(animated: true)
+        })
+//        defer {
```
This code is commented out, you planning on doing something with it?

nathantannar4 replied:
Once this gets merged I will be updating the entire example app with more detailed examples for the TypingIndicator and custom cells.

SD10 commented on Sources/Controllers/MessagesViewController.swift:
```diff
@@ -147,20 +167,79 @@ UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
+    ///   - updates: A block of code that will be executed during `performBatchUpdates`
+    ///              when `animated` is `TRUE` or before the `completion` block executes
+    ///              when `animated` is `FALSE`
+    ///   - completion: A completion block to execute after the insertion/deletion
```
Very nice docs 💯 🥇

SD10 commented on CHANGELOG.md:
```diff
@@ -8,6 +8,10 @@ The changelog for `MessageKit`. Also see the [releases](https://github.com/Messa
 
 ### Added
 
+- **Breaking Change** Added typing indicator support to MessageKit by adding an extra section to `MessagesCollectionView` when needed. This included adding the following views: `BubbleCircle`, `TypingBubble`, `TypingIndicator`, `TypingBubbleCell`. Determine if the section is reserved for the typing indicator with `isSectionReservedForTypingIndicator(_ section: Int) -> Bool`, and set the typing in ...
```
This is more a reminder to myself to improve this CHANGELOG entry before merging

zhongwuzw replied:
I don't know how we can tell the user that `numberOfSections(in collectionView: UICollectionView)` would not only return the data source the user provide.

SD10 replied:
Yes, this is a valid concern that I have as well. I like the general idea but I haven't gone deep into how this will affect MessageKit

stasian23 commented:
Hi guys, are you planning to merge it to dev, and release? We really need this feature in our project

SD10 commented:
@stasian23 That's the plan but this is lower priority right now. I like the general idea but I haven't thoroughly reviewed the code to see what impact this PR would have

nathantannar4 commented:
@stasian23 I just rebased the`typingBubble` branch onto the `development` branch. You can specify in your pod file to pull from the `typingBubble` branch if you would like to use it in pre-release

zhongwuzw commented on Sources/Controllers/MessagesViewController.swift:
```diff
@@ -163,20 +183,79 @@ UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
+    ///              when `animated` is `TRUE` or before the `completion` block executes
+    ///              when `animated` is `FALSE`
+    ///   - completion: A completion block to execute after the insertion/deletion
+    open func setTypingIndicatorHidden(_ isHidden: Bool, animated: Bool, whilePerforming updates: (() -> Void)? = nil, completion: ((Bool) -> Void)?=nil) {
```
`completion: ((Bool) -> Void)?=nil`,please keep indention between nil.

zhongwuzw commented on Sources/Controllers/MessagesViewController.swift:
```diff
@@ -163,20 +183,79 @@ UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
+        let section = messagesCollectionView.numberOfSections
+        
+        isTypingIndicatorHidden = isHidden
+        if let messagesFlowLayout = messagesCollectionView.collectionViewLayout as? MessagesCollectionViewFlowLayout {
```
Maybe `guard return` would be better, if `layout` is not `MessagesCollectionViewFlowLayout`, nothing we can do.

zhongwuzw commented on Sources/Controllers/MessagesViewController.swift:
```diff
@@ -185,6 +264,14 @@ UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
+            cell.typingBubble.typingIndicator.dotColor = typingBubbleDotColor
+            cell.typingBubble.startAnimating()
+            return cell
+        }
```
Just a suggestion, can we move these things to a separate function? `collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath)` becomes the big function.

SD10 replied:
Yeah, it should be a function on the cell

zhongwuzw commented on Sources/Views/Cells/TypingBubbleCell.swift:
```diff
@@ -0,0 +1,66 @@
+    }
+    
+    open func setupSubviews() {
+        contentView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
```
Is this necessary?

nathantannar4 replied:
Thats how other `MessageContentCell` are set up so I added this for consistency

SD10 replied:
Don't copy my code if it's bad (it usually is) 😆

zhongwuzw commented on Sources/Views/Cells/TypingBubbleCell.swift:
```diff
@@ -0,0 +1,66 @@
+        super.prepareForReuse()
+        if typingBubble.isAnimating {
+            typingBubble.stopAnimating()
+            typingBubble.startAnimating()
```
`startAnimating()` is not necessary IMO.

nathantannar4 replied:
The animations stop when they move off screen, so they need to be restarted when they are about to appear again

zhongwuzw replied:
@nathantannar4 May I miss something? you already start it in `cellForItemAt`. 🤔

nathantannar4 replied:
Its required for the cell reusability. If its not initially started in cellForItem then isAnimating will be false

nathantannar4 committed:
Cleanup and added `dequeuedTypingCell(for indexPath: IndexPath) `

`dequeuedTypingCell(for indexPath: IndexPath) ` can be used for easily overriding the typing cell and keeps `collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath)` clean

nathantannar4 commented:
Recommendations added

nathantannar4 committed:
Rebase onto `development`

gorbat-o commented:
Hey,

Is there any progress on this PR ? 

I hope it can be merged soon :D

nathantannar4 commented:
2.0 release 😄

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

gorbat-o commented:
😱😱😱😱😱
When  ??

nathantannar4 commented:
@SD10 and I have had some offline discussion about this and how this method reserves the last section of the `MessagesCollectionView`.

@SD10 I know you mentioned reserving the last section would be an issue for RxSwift users, but is there anything else?

nathantannar4 committed:
Update to Swift 4.2

nathantannar4 commented:
Re-evaluating this PR so that is does not reserve a section

nathantannar4 committed:
Merge branch 'development' into typingBubble

nathantannar4 commented:
@SD10 @zhongwuzw I am revisiting this so we can get it merged. Are there still hesitations to adding an additional section? I tried using supplementary views but it becomes much more challenging to manage since the indicator ends up needing to be tied to a specific section. There are just many more edge cases that route.

SD10 commented:
> Are there still hesitations to adding an additional section?

It's been so long I'll have to re-evaluate reserving the last section for the typing indicator but I do think the highly dynamic nature of a typing indicator would cause a lot of `IndexPath` inconsistencies with apps in a production environment.

Probably not a good candidate for 3.0

zhongwuzw commented:
I agree with @SD10 , can we put the indicator view as a subview of `UICollectionView`? not treat it as a cell.
